// editor/pegjs-parser/helpers/node-instancier/index.js
var oop = require('oop').oop;

var node = require('editor/node/node');





var NodeInstancier = oop.Class({
	name: 'Node Instancier',

	init: function(language, light_node) {
		if (light_node == null) {
			light_node = false;
		}

		this.language = language != null ? language : null;
		this.NodeClass = light_node ? node.light.Node : node.complete.Node;

		this.hooks = [];
	},

	call: function(element, line, column, start, source) {
		// ---------------------------------------------------------------------

		var node = new this.NodeClass({
			type: {
				language: this.language,
				element: element
			},
			location: {
				start: {
					line: line,
					column: column,
					index: start
				},
				end: {
					index: start + source.length
				}
			},
			source: source
		});

		// ---------------------------------------------------------------------

		var hooks = this.hooks;
		for (var i = 0, length = hooks.length; i < length; i++) {
			hooks[i](node);
		}

		// ---------------------------------------------------------------------

		return node;
	},

	methods: {
		addHook: function(fn) {
			this.hooks.push(fn);
			return fn;
		}
	}
});





exports.NodeInstancier = NodeInstancier;
