// editor/highlight/color/index.js
var prelude = require('prelude-ls');
var oop = require('oop').oop;





function ensureRange(min, max, value) {
	var result = value;
	if (value < min) result = min;
	else if (value > max) result = max;
	return result;
}





function postprocessColor(value) {
	return ensureRange(0, 255, value);
}



var Color = oop.Class({
	name: 'Color',

	desc: 'Represents a color as RGB.',

	schema: {
		inputToSpec: {
			'String': function(input) {
				var color = Color.defaults[input];

				if (color != null) return color;
				throw {
					msg: 'No default color under this name',
					name: input
				};
			},
			'Number': function(input) {
				return Color.greyScale(input);
			},
			'Array': function(input) {
				return Color({
					r: input[0],
					g: input[1],
					b: input[2]
				})
			}
		},
		properties: [
			{names: ['r', 'R', 'red'], type: oop.types.Number, postprocess: postprocessColor},
			{names: ['g', 'G', 'green'], type: oop.types.Number, postprocess: postprocessColor},
			{names: ['b', 'B', 'blue'], type: oop.types.Number, postprocess: postprocessColor}/*,

			{names: ['h', 'H', 'hue'], type: oop.types.Number},
			{names: ['s', 'S', 'saturation'], type: oop.types.Number},

			{names: ['l', 'L', 'lightness'], type: oop.types.Number},
			{names: ['v', 'V', 'value'], type: oop.types.Number},
			{names: ['i', 'I', 'intensity'], type: oop.types.Number}
			*/
		]
	},

	methods: {
		serialize: function() {
			return "rgb(" + this.r + ", " + this.g + ", " + this.b + ")";
		},

		css: function() {
			return this.serialize();
		}
	},

	statics: {
		parse: function(input) {
			var components = /rgb\(\s*([^,]+),\s*([^,]+),\s*([^,]+)\)/.exec(input);
			return new Color({
				r: components[1],
				g: components[2],
				b: components[3]
			});
		}
	},

 	factories: {
 		rgb: {
 			args: {r: 0, g: 1, b: 2}
 		},
 		greyScale: {
 			args: {r: 0, g: 0, b: 0}
 		},
 		redScale: {
 			args: {r: 0},
 			values: {g: 0, b: 0}
 		},
 		blueScale: {
 			args: {b: 0},
 			values: {r: 0, g: 0}
 		},
 		greenScale: {
 			args: {g: 0},
 			values: {r: 0, b: 0}
 		}
 	}
});



oop.addStatics(Color, {
	defaults: {
		// Grey scales ---------------------------------------------------------

		'black': Color.greyScale(0),
		'gray': Color.greyScale(128),
		'silver': Color.greyScale(192),
		'white': Color.greyScale(255),

		// Plain colors --------------------------------------------------------

		'red': Color.redScale(255),
		'lime': Color.greenScale(255),
		'blue': Color.blueScale(255),

		'yellow': Color.rgb(255, 255, 0),
		'aqua': Color.rgb(0, 255, 255),
		'fushia': Color.rgb(255, 0, 255),

		// Others --------------------------------------------------------------

		'purple': Color.rgb(128, 0, 128),
		'maroon': Color.redScale(128),
		'olive': Color.rgb(128, 128, 0),
		'teal': Color.rgb(0, 128, 128),
		'navy': Color.blueScale(128),
	}
});

// Add color aliases

var aliases = {
	'gray': ['grey', 'dark gray', 'dark grey'],

	'red': ['high red'],
	'maroon': ['low red'],

	'olive': ['brown'],

	'lime': ['high green'],
	'green': ['low green'],

	'aqua': ['high cyan'],
	'teal': ['low cyan'],

	'blue': ['high blue'],
	'navy': ['low blue'],

	'fushia': ['high magenta'],
	'purple': ['low magenta']
}

for (var colorName in aliases) {
	var color = Color.defaults[colorName];

	var alternatives = aliases[colorName];
	for (var i = 0, length = alternatives.length; i < length; i++) {
		Color.defaults[alternatives[i]] = color;
	}
}


exports.Color = Color;
